src/cargo/ops/cargo_new.rs

```diff
@@ -44,38 +58,235 @@ struct CargoNewConfig {
     version_control: Option<VersionControl>,
 }
 
-pub fn new(opts: NewOptions, config: &Config) -> CargoResult<()> {
-    let path = config.cwd().join(opts.path);
-    if fs::metadata(&path).is_ok() {
-        bail!("destination `{}` already exists", path.display())
+fn get_name<'a>(path: &'a Path, opts: &'a NewOptions, config: &Config) -> CargoResult<&'a str> {
+    if let Some(name) = opts.name {
+        return Ok(name);
     }
-    let name = match opts.name {
-        Some(name) => name,
-        None => {
-            let dir_name = try!(path.file_name().and_then(|s| s.to_str()).chain_error(|| {
-                human(&format!("cannot create a project with a non-unicode name: {:?}",
-                               path.file_name().unwrap()))
-            }));
-            if opts.bin {
-                dir_name
-            } else {
-                let new_name = strip_rust_affixes(dir_name);
-                if new_name != dir_name {
-                    let message = format!(
-                        "note: package will be named `{}`; use --name to override",
-                        new_name);
-                    try!(config.shell().say(&message, BLACK));
-                }
-                new_name
-            }
+    
+    if path.file_name().is_none() {
+        bail!("cannot auto-detect project name from path {:?} ; use --name to override",
+                              path.as_os_str());
+    }
+    
+    let dir_name = try!(path.file_name().and_then(|s| s.to_str()).chain_error(|| {
+        human(&format!("cannot create a project with a non-unicode name: {:?}",
+                       path.file_name().unwrap()))
+    }));
+    
+    if opts.bin {
+        Ok(dir_name)
+    } else {
+        let new_name = strip_rust_affixes(dir_name);
+        if new_name != dir_name {
+            let message = format!(
+                "note: package will be named `{}`; use --name to override",
+                new_name);
+            try!(config.shell().say(&message, BLACK));
         }
-    };
+        Ok(new_name)
+    }
+}
+
+fn check_name(name: &str) -> CargoResult<()> {
     for c in name.chars() {
         if c.is_alphanumeric() { continue }
         if c == '_' || c == '-' { continue }
-        bail!("Invalid character `{}` in crate name: `{}`", c, name)
+        bail!("Invalid character `{}` in crate name: `{}`
```

Comment:
I think this will need a trailing `\` character to ensure that the next line doesn't start with a bunch of whitespace